File: src/test/java/leets/weeth/domain/attendance/domain/service/AttendanceUpdateServiceTest.java
package leets.weeth.domain.attendance.domain.service;

import static leets.weeth.domain.attendance.test.fixture.AttendanceTestFixture.*;
import static leets.weeth.domain.schedule.test.fixture.ScheduleTestFixture.*;
import static org.mockito.Mockito.*;

import java.util.List;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import leets.weeth.domain.attendance.domain.entity.Attendance;
import leets.weeth.domain.attendance.domain.entity.enums.Status;
import leets.weeth.domain.schedule.domain.entity.Meeting;
import leets.weeth.domain.user.domain.entity.User;

class AttendanceUpdateServiceTest {

	private final AttendanceUpdateService attendanceUpdateService = new AttendanceUpdateService();

	@Test
	@DisplayName("attend(): attendance.attend() + user.attend() 호출")
	void attend_callsEntityMethods() {
		// given
		Meeting meeting = createMeeting();
		User realUser = createActiveUser("이지훈");
		User userSpy = spy(realUser);

		doNothing().when(userSpy).attend();

		Attendance realAttendance = createAttendance(meeting, userSpy);
		Attendance attendanceSpy = spy(realAttendance);

		// when
		attendanceUpdateService.attend(attendanceSpy);

		// then
		verify(attendanceSpy).attend();
		verify(userSpy).attend();
	}

	@Test
	@DisplayName("close(): pending만 close() + user.absent() 호출")
	void close_onlyPendingIsClosed() {
		// given
		Meeting meeting = createMeeting();

		User pendingUserReal = createActiveUser("pending-user");
		User nonPendingUserReal = createActiveUser("non-pending-user");
		User pendingUserSpy = spy(pendingUserReal);
		User nonPendingUserSpy = spy(nonPendingUserReal);

		doNothing().when(pendingUserSpy).absent();
		doNothing().when(nonPendingUserSpy).absent();

		Attendance pendingAttendanceReal = createAttendance(meeting, pendingUserSpy);
		Attendance nonPendingAttendanceReal = createAttendance(meeting, nonPendingUserSpy);
		Attendance pendingAttendanceSpy = spy(pendingAttendanceReal);
		Attendance nonPendingAttendanceSpy = spy(nonPendingAttendanceReal);

		doReturn(true).when(pendingAttendanceSpy).isPending();
		doReturn(false).when(nonPendingAttendanceSpy).isPending();

		// when
		attendanceUpdateService.close(List.of(pendingAttendanceSpy, nonPendingAttendanceSpy));

		// then
		verify(pendingAttendanceSpy).close();
		verify(pendingUserSpy).absent();

		verify(nonPendingAttendanceSpy, never()).close();
		verify(nonPendingUserSpy, never()).absent();
	}

	@Test
	@DisplayName("updateUserAttendanceByStatus: ATTEND면 user.removeAttend(), 그 외에는 user.removeAbsent()")
	void updateUserAttendanceByStatus() {
		// given
		Meeting meeting = createMeeting();

		User attendUserReal = createActiveUser("attend-user");
		User absentUserReal = createActiveUser("absent-user");

		User attendUserSpy = spy(attendUserReal);
		User absentUserSpy = spy(absentUserReal);

		doNothing().when(attendUserSpy).removeAttend();
		doNothing().when(absentUserSpy).removeAbsent();

		Attendance attendAttendanceReal = createAttendance(meeting, attendUserSpy);
		Attendance absentAttendanceReal = createAttendance(meeting, absentUserSpy);
		Attendance attendAttendanceSpy = spy(attendAttendanceReal);
		Attendance absentAttendanceSpy = spy(absentAttendanceReal);

		doReturn(Status.ATTEND).when(attendAttendanceSpy).getStatus();
		doReturn(Status.ABSENT).when(absentAttendanceSpy).getStatus();
		doReturn(attendUserSpy).when(attendAttendanceSpy).getUser();
		doReturn(absentUserSpy).when(absentAttendanceSpy).getUser();

		// when
		attendanceUpdateService.updateUserAttendanceByStatus(List.of(attendAttendanceSpy, absentAttendanceSpy));

		// then
		verify(attendUserSpy).removeAttend();
		verify(absentUserSpy).removeAbsent();
	}
}

File: src/test/kotlin/com/weeth/domain/attendance/domain/service/AttendanceUpdateServiceTest.kt
package com.weeth.domain.attendance.domain.service

import com.weeth.domain.attendance.domain.entity.enums.AttendanceStatus
import com.weeth.domain.attendance.fixture.AttendanceTestFixture.createActiveUser
import com.weeth.domain.attendance.fixture.AttendanceTestFixture.createAttendance
import com.weeth.domain.attendance.fixture.AttendanceTestFixture.createOneDaySession
import io.kotest.core.spec.style.DescribeSpec
import io.mockk.every
import io.mockk.spyk
import io.mockk.verify
import java.time.LocalDate

class AttendanceUpdateServiceTest :
    DescribeSpec({

        val attendanceUpdateService = AttendanceUpdateService()

        describe("attend") {
            it("attendance.attend() + user.attend()을 호출한다") {
                val session = createOneDaySession(LocalDate.now(), 1, 1234, "1차")
                val userSpy = spyk(createActiveUser("이지훈"))
                every { userSpy.attend() } returns Unit

                val attendanceSpy = spyk(createAttendance(session, userSpy))

                attendanceUpdateService.attend(attendanceSpy)

                verify { attendanceSpy.attend() }
                verify { userSpy.attend() }
            }
        }

        describe("close") {
            it("pending만 close() + user.absent()을 호출한다") {
                val session = createOneDaySession(LocalDate.now(), 1, 1234, "1차")

                val pendingUserSpy = spyk(createActiveUser("pending-user"))
                val nonPendingUserSpy = spyk(createActiveUser("non-pending-user"))
                every { pendingUserSpy.absent() } returns Unit
                every { nonPendingUserSpy.absent() } returns Unit

                val pendingAttendanceSpy = spyk(createAttendance(session, pendingUserSpy))
                val nonPendingAttendanceSpy = spyk(createAttendance(session, nonPendingUserSpy))
                every { pendingAttendanceSpy.isPending() } returns true
                every { nonPendingAttendanceSpy.isPending() } returns false

                attendanceUpdateService.close(listOf(pendingAttendanceSpy, nonPendingAttendanceSpy))

                verify { pendingAttendanceSpy.close() }
                verify { pendingUserSpy.absent() }

                verify(exactly = 0) { nonPendingAttendanceSpy.close() }
                verify(exactly = 0) { nonPendingUserSpy.absent() }
            }
        }

        describe("updateUserAttendanceByStatus") {
            it("ATTEND면 user.removeAttend(), 그 외에는 user.removeAbsent()") {
                val session = createOneDaySession(LocalDate.now(), 1, 1234, "1차")

                val attendUserSpy = spyk(createActiveUser("attend-user"))
                val absentUserSpy = spyk(createActiveUser("absent-user"))
                every { attendUserSpy.removeAttend() } returns Unit
                every { absentUserSpy.removeAbsent() } returns Unit

                val attendAttendanceSpy = spyk(createAttendance(session, attendUserSpy))
                val absentAttendanceSpy = spyk(createAttendance(session, absentUserSpy))
                every { attendAttendanceSpy.status } returns AttendanceStatus.ATTEND
                every { absentAttendanceSpy.status } returns AttendanceStatus.ABSENT
                every { attendAttendanceSpy.user } returns attendUserSpy
                every { absentAttendanceSpy.user } returns absentUserSpy

                attendanceUpdateService.updateUserAttendanceByStatus(
                    listOf(attendAttendanceSpy, absentAttendanceSpy),
                )

                verify { attendUserSpy.removeAttend() }
                verify { absentUserSpy.removeAbsent() }
            }
        }
    })
